### PR TITLE
MM-42655 Improve memoization of permission selectors

### PR DESCRIPTION
**Pull request by hmhealey**
I'm going to talk about the channel permissions selectors below, but the same applies to the ones used for teams and groups as well.

The previous versions of `haveIChannelPermission` was memoized with createSelector, but because it takes arguments, two components calling `haveIChannelPermission` for different channels would pave over each others' results, possibly multiple times in a single render cycle. I tried just removing that, but the internal selectors it used to generate a set of all permissions for a given channel also had the same problem.

Instead of just removing all memoization which would've just gave us the same result (recalculating everything constantly), I changed it instead so that we generate a map of permissions for each channel and use that in `haveIChannelPermission`. Because permissions cascade, this does mean we have to check multiple sets of permissions (as opposed to before where all permissions were in one big set), but that should have a negligible effect on performance.

I also considered using a selector factory for `haveIChannelPermission` or `getMyChannelPermissions` instead of what I ended up with, but that seemed way more complicated and would require changes in a lot more files.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42655

#### Release Note
```release-note
Improved performance of permission selectors
```

### CODE REVIEW

**Review by BenCookie95 — APPROVED** [review]
LGTM, nice work!

File: packages/mattermost-redux/src/selectors/entities/roles.ts [hunk 1]
```diff
@@ -141,179 +124,77 @@ export const getRolesById: (state: GlobalState) => Record<string, Role> = create
     },
 );
 
-export const getMyCurrentTeamPermissions: (state: GlobalState) => Set<string> = createSelector(
-    'getMyCurrentTeamPermissions',
+const getMyPermissionsByTeam = createSelector(
```
[the remark this answers is not in the document]
Comment (jespino) [reply packages/mattermost-redux/src/selectors/entities/roles.ts:127]: Nevermind, I now understand what is going on.

**Review by jespino — APPROVED** [review]
LGTM

**Review by jgilliam17 — APPROVED** [review]

**Pull request merged**